doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,302 @@
+stream, the errs stream is always presented as a relation with two columns:
+
+ Name      | Type
+-----------|--------
+ `message` | `text`
+ `code`    | `text`
+
+
+### Guardrails
+
+#### `IGNORE ERRORS`
+
+As mentioned above, the `IGNORE ERRORS` clause can have dangerous consequences
+for correctness. Materialize makes no general claims about the contents of the
+oks stream when errors are present in the errs stream.
+
+However, for certain simple types of computations, we can offer better
+guarantees. Specifically, for sources, we can guarantee that an error while
+decoding a record will only cause that one record to be omitted from the oks
+stream.
+
+So, to start, we will permit applying the `IGNORE ERRORS` modifier only to
+sources and subsources.
+
+In the future, we can consider expanding the set of relations to which `IGNORE
+ERRORS` can be applied. For example, applying `IGNORE ERRORS` to a SQL query
+which contains only map, filter, and project nodes in its `HirScalarExpr` likely
+has sensible semantics under `IGNORE ERRORS`: the computation proceeds row by
+row, and any errors while applying the computation to a row omits only that one
+row from the output.
+
+#### `ONLY ERRORS`
+
+There are two important guardrails to apply to `ONLY ERRORS`:
+
+1. We will permit applying the `ONLY ERRORS` modifier only to persist-backed
+   objects: tables, sources, and materialized views. This relieves some pressure
+   on the compute layer, which already does not assume that the errors in a
+   persist shard have any particular relationship to their source.
+
+2. We will clearly document that error messages and codes are **not stable**. We
+   reserve the right to reword error messages, change error codes, add new
+   errors, or eliminate errors in future releases of Materialize.
+
+   Note also that `ONLY ERRORS` doesn't fundamentally add new backwards
+   compatibility surface area. Error messages, descriptions, hints, and codes
+   are already returned via the SQL interface, and users may already be
+   erroneously relying on seeing particular error messages and codes from
+   Materialize in particular situations.
+
+### Implementation
+
+#### `IGNORE ERRORS`
+
+[@maddyblue] has a prototype implementation of `IGNORE ERRORS` in [#27137].
+The core of the implementation is the addition of a new `ignore_errors`
+field to `MirRelationExpr::Get`:
+
+```diff
+ MirRelationExpr::Get {
+     /// A global or local identifier naming the collection.
+     id: Id,
+
+     // ...
+
++    /// Whether to ignore errors.
++    ignore_errors: bool,
+ },
+```
+
+When rendered, if `ignore_errors` is true, the referenced collection's errs
+stream is simply not wired up to the downstream error stream.
+
+**Open questions:**
+
+  * Is this implementation acceptable to the cluster team?
+  * Will the existence of the new field cause correctness issues if existing
+    references to `MirRelationExpr` fail to check the field?
```

Comment:
> I'd say we should disallow this in `LocalIds`, so that MIR manipulations don't need to take this into account in complicated ways, but just have to make sure that this doesn't fall off when moving around `Get`s.

Perhaps that is an argument for the `ignore_errors` field being part of `Id::Global` rather than `MirRelationExpr::Get`—i.e., as I sketched out below for `ONLY ERRORS`.